### PR TITLE
Introduce compatibility layer for collective operations

### LINKED ISSUE #5395: HPX 1.7.0-rc1 fails to build icw APEX + OTF2
## Expected Behavior

Build succeeds.

## Actual Behavior

Build fails with these messages (full log attached):

```
apex-src/src/apex/otf2_listener.cpp:1402:59: error: ‘gather_there’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_there’?
apex-src/src/apex/otf2_listener.cpp:1412:28: error: ‘gather_here’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_here’?
apex-src/src/apex/otf2_listener.cpp:1455:59: error: ‘gather_there’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_there’? apex-src/src/apex/otf2_listener.cpp:1464:32: error: ‘gather_here’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_here’?
apex-src/src/apex/otf2_listener.cpp:1526:32: error: ‘broadcast_from’ is not a member of ‘hpx::lcos’; did you mean ‘broadcast_apply’?
apex-src/src/apex/otf2_listener.cpp:1526:58: error: expected primary-expression before ‘>’ token               
apex-src/src/apex/otf2_listener.cpp:1531:32: error: ‘broadcast_to’ is not a member of ‘hpx::lcos’; did you mean ‘broadcast’?
apex-src/src/apex/otf2_listener.cpp:1585:59: error: ‘gather_there’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_there’?
apex-src/src/apex/otf2_listener.cpp:1594:32: error: ‘gather_here’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_here’?
apex-src/src/apex/otf2_listener.cpp:1662:32: error: ‘broadcast_from’ is not a member of ‘hpx::lcos’; did you mean ‘broadcast_apply’?
apex-src/src/apex/otf2_listener.cpp:1662:58: error: expected primary-expression before ‘>’ token               
apex-src/src/apex/otf2_listener.cpp:1667:32: error: ‘broadcast_to’ is not a member of ‘hpx::lcos’; did you mean ‘broadcast’?
apex-src/src/apex/otf2_listener.cpp:1714:59: error: ‘gather_there’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_there’?
apex-src/src/apex/otf2_listener.cpp:1724:28: error: ‘gather_here’ is not a member of ‘hpx::lcos’; did you mean ‘hpx::collectives::gather_here’?
```

## Steps to Reproduce the Problem

  1. Configure HPX with `-DHPX_WITH_APEX:BOOL=ON -DAPEX_WITH_OTF2:BOOL=ON -DHELLO_HPX_BUILD_OTF2:BOOL=ON -DHPX_WITH_EXAMPLES:BOOL=ON`
  
I test HPX builds using this project: https://github.com/kordejong/hello_hpx The build that fails is the one in https://github.com/kordejong/hello_hpx/blob/b8f53591c9b6fcfcc6e5ce887bd48c23ffe2c77d/rel_with_deb_info_build.sh#L12

## Specifications

  - HPX Version: 1.7.0-rc1
  - Platform (compiler, OS): gcc-9.3.0, Linux

[build_log.txt](https://github.com/STEllAR-GROUP/hpx/files/6662989/build_log.txt)

### PR DESCRIPTION
**Pull request by hkaiser**
- flyby: updating documentation for collective operations

Fixes #5395

### CODE REVIEW

File: libs/full/collectives/include/hpx/collectives/all_reduce.hpp [hunk 1]
```diff
@@ -219,6 +219,87 @@ namespace hpx { namespace collectives {
...
+        "hpx::lcos::all_reduce is deprecated, use hpx::collectives::all_reduce "
+        "instead")
+    hpx::future<T> all_reduce(char const* basename,
+        hpx::future<T>&& local_result, F&& op,
+        std::size_t num_sites = std::size_t(-1),
+        std::size_t generation = std::size_t(-1),
+        std::size_t this_site = std::size_t(-1), std::size_t root_site = 0)
+    {
+        return local_result.then([=](hpx::future<T>&& f) {
+            return hpx::collectives::all_reduce(basename, f.get(), std::move(op),
```
Comment (msimberg) [inline libs/full/collectives/include/hpx/collectives/all_reduce.hpp:254]:
```suggestion
            return hpx::collectives::all_reduce(basename, f.get(), std::forward<F>(op),
```
and others, as pointed out by clang-tidy.

**Review by msimberg — APPROVED** [review]
Besides the clang-tidy warnings, looks good to me!

File: libs/full/collectives/include/hpx/collectives/all_reduce.hpp (hunk 1 above)
Comment (hkaiser) [reply libs/full/collectives/include/hpx/collectives/all_reduce.hpp:254]: I think, the `move` is correct (op was captured by value). The lambda has to be `mutable`, though.

File: libs/full/collectives/include/hpx/collectives/all_reduce.hpp [hunk 2]
```diff
@@ -219,6 +219,91 @@ namespace hpx { namespace collectives {
...
+        "instead")
+    hpx::future<T> all_reduce(char const* basename,
+        hpx::future<T>&& local_result, F&& op,
+        std::size_t num_sites = std::size_t(-1),
+        std::size_t generation = std::size_t(-1),
+        std::size_t this_site = std::size_t(-1), std::size_t root_site = 0)
+    {
+        return local_result.then([=](hpx::future<T>&& f) mutable {
+            // silence bogus clang-tidy error
+            // NOLINTNEXTLINE(bugprone-move-forwarding-reference)
```
Comment (msimberg) [inline libs/full/collectives/include/hpx/collectives/all_reduce.hpp:255]: The comment needs to move one line further down.

**Review by biddisco — APPROVED** [review]
Seems like a huge waste of time to add a compatibility layer to something only you are using!
Approved, subject to the comments Mikael already made

**Comment by hkaiser** [comment]
> Seems like a huge waste of time to add a compatibility layer to something only you are using!

I thought so too, but as it turned out, two people started to complain about the breaking change.

**Pull request merged**